File: api_python/examples/103-Gen3_uart_bridge/01-uart_bridge.py
#! /usr/bin/env python3

###
# KINOVA (R) KORTEX (TM)
#
# Copyright (c) 2019 Kinova inc. All rights reserved.
#
# This software may be modified and distributed under the
# terms of the BSD 3-Clause license.
#
# Refer to the LICENSE file for details.
#
###

###
# 103-Gen3_uart_bridge/uart_bridge.py
#
# Usage example for the UART Expansion IO on the Gen3 Interconnect module.
#
# PHYSICAL SETUP:
# =========
# For this example to work, you will have to connect the RX and TX pins on the interconnect 
# IO Expansion pinout, so that the TX is transmitted right into the RX.
#
# DESCRIPTION OF CURRENT EXAMPLE:
# ===============================
# In this example, the UART bridge class encapsulates all necessary Kortex API
# objects and implements the functions to setup the UART, write to the bus and read back from it.
# 
# The Init function creates the Kortex API objects, connects to the arm and also calls InitSocket, 
# which correctly configures the socket (depending if you are on Windows or Linux).
#
# The Run function :
#   - Tells the Interconnect (m_interconnect_config object) to enable the UART bridge (this is where you supply the UART parameters)
#   - Tells the Base (m_base) to enable the UART bridge
#   - Opens the TCP socket to the base and writes the test string to the UART bridge
#   - Waits 10 seconds and prints every character it receives
#   - Tells the Base (m_base) to disable the UART bridge
#   - Tells the Interconenct (m_interconnect_config_object) to disable the UART bridge
###

import sys
import time
import os
import socket
import select

from kortex_api.autogen.client_stubs.BaseClientRpc import BaseClient
from kortex_api.autogen.client_stubs.DeviceManagerClientRpc import DeviceManagerClient
from kortex_api.autogen.client_stubs.InterconnectConfigClientRpc import InterconnectConfigClient

from kortex_api.autogen.messages import Base_pb2
from kortex_api.autogen.messages import Common_pb2
from kortex_api.autogen.messages import InterconnectConfig_pb2

class UARTBridge:

    def __init__(self, router, ip_address):

        self.router = router
        self.base_ip_address = ip_address

        # Create services
        self.base = BaseClient(self.router)
        self.device_manager =  DeviceManagerClient(self.router)
        self.interconnect_config = InterconnectConfigClient(self.router)

        self.interconnect_device_id = self.GetDeviceIdFromDevType(Common_pb2.INTERCONNECT, 0)
        if (self.interconnect_device_id is None):
            print ("Could not find the Interconnect in the device list, exiting...")
            sys.exit(0)

    def GetDeviceIdFromDevType(self, device_type, device_index = 0):
        devices = self.device_manager.ReadAllDevices()

        current_index = 0
        for device in devices.device_handle:
            if device.device_type == device_type:
                if current_index == device_index:
                    print ("Found the Interconnect on device identifier {}".format(device.device_identifier))
                    return device.device_identifier
                current_index += 1
        return None

    def Configure(self, port_id, enabled, speed, word_length, stop_bits, parity):
        '''
        Enable and configure UART on interconnect. This will open a TCP port on the interconnect. This
        port allows bridging TCP socket to UART.
        '''
        uart_config              = Common_pb2.UARTConfiguration()
        uart_config.port_id      = port_id 
        uart_config.enabled      = enabled # Setting enabled to true opens the TCP port dedicated to UART bridging. Setting this
                                           # field to false disables designated uart and closes the TCP port.
        uart_config.speed        = speed 
        uart_config.word_length  = word_length
        uart_config.stop_bits    = stop_bits
        uart_config.parity       = parity

        self.interconnect_config.SetUARTConfiguration(uart_config, deviceId=self.interconnect_device_id)

    def EnableBridge(self, bridge_type, target = 0, output = 0):
        
        # Create bridge configuration
        bridge_config = Base_pb2.BridgeConfig()
        bridge_config.device_identifier = self.interconnect_device_id
        bridge_config.bridgetype = bridge_type

        # If either target or ouput port has valid port value, add port config to bridge configuration
        if target or output:
            bridge_config.port_config.target_port = 0
            bridge_config.port_config.out_port = 0
            if target:
                bridge_config.port_config.target_port = target
            if output:
                bridge_config.port_config.out_port = output

        # Send the configuration and return the result
        bridge_result = self.base.EnableBridge(bridge_config)
        return bridge_result

    def DisableBridge(self, bridge_id):
        return self.base.DisableBridge(bridge_id)

    def ExampleSendDataAndReadItBack(self):

        # Enable port bridging on base.
        bridge_result = self.EnableBridge(Base_pb2.BRIDGE_TYPE_UART)
        bridge_id = bridge_result.bridge_id
        if bridge_result.status != Base_pb2.BRIDGE_STATUS_OK:
            print("Error creating bridge on base, exiting...")
            return

        # Get created bridge's configuration.
        bridge_config       = self.base.GetBridgeConfig(bridge_id)
        base_port           = bridge_config.port_config.out_port
        interconnect_port   = bridge_config.port_config.target_port
        print("UARTBridge ID # %i created between Interconnect (dev# %i)'s port #%i and external port #%i"%(bridge_id.bridge_id, self.interconnect_device_id, interconnect_port, base_port))

        # Open a socket to base's forwarded port.
        client_socket =  socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        client_socket.connect((self.base_ip_address, base_port))
        client_socket.setblocking(0)

        print ("UART bridge object initialized")

        # Send data to be written on UART using TCP connection.
        client_socket.send(b"This data is being written on Interconnect's expansion port UART\n")

        # Wait for data to be received from UART
        print("Waiting 10 seconds for data from uart...")
        sys.stdout.write("Received data : ")
        sys.stdout.flush()
        startTime = time.time()
        while time.time() - startTime < 10:
            readready, _, _  = select.select([client_socket], [], [], 1)
            if readready:
                data = client_socket.recv(1)
                if len(data):
                    sys.stdout.write(data.decode("utf-8"))
                    sys.stdout.flush()

        # Disconnect client socket.
        client_socket.close()

        # Disable bridge on base.
        self.DisableBridge(bridge_id)

def main():

    # Import the utilities helper module
    import argparse
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))
    import utilities

    # Parse arguments
    parser = argparse.ArgumentParser()
    args = utilities.parseConnectionArguments(parser)

    # Create connection to the device and get the router
    with utilities.DeviceConnection.createTcpConnection(args) as router:

        # Create the bridge object. UARTBridge object implements kortex methods used
        # to create a TCP port bridge (port forwarding) between a port opened
        # on external adapter and the port opened on the interconnect.
        bridge = UARTBridge(router, args.ip)

        # Configure the expansion IO
        bridge.Configure(InterconnectConfig_pb2.UART_PORT_EXPANSION, True, Common_pb2.UART_SPEED_115200, Common_pb2.UART_WORD_LENGTH_8, Common_pb2.UART_STOP_BITS_1, Common_pb2.UART_PARITY_NONE)
        time.sleep(1)

        # Example core        
        bridge.ExampleSendDataAndReadItBack()

        # Disable bridge on interconnect
        bridge.Configure(InterconnectConfig_pb2.UART_PORT_EXPANSION, False, Common_pb2.UART_SPEED_115200, Common_pb2.UART_WORD_LENGTH_8, Common_pb2.UART_STOP_BITS_1, Common_pb2.UART_PARITY_NONE)
        
"""
This example will configure uart and activate uart to TCP socket bridge  on the arm's interconnect. It will also activate
TCP port bridge (port forwarding) to allow accessing the opened TCP port on the interconnect from outside the arm.
Once uart bridge is configured, it will send a string to write on UART port and then will wait for 10 seconds for data to
be entered on the same serial port. Whenever a character is received from UART it is written on stdout.
"""
if __name__ == "__main__":
    main()

File: api_cpp/examples/103-Gen3_uart_bridge/01-uart_bridge.cpp
/*
* KINOVA (R) KORTEX (TM)
*
* Copyright (c) 2019 Kinova inc. All rights reserved.
*
* This software may be modified and distributed
* under the terms of the BSD 3-Clause license.
*
* Refer to the LICENSE file for details.
*
*/

/*
* 103-Gen3_uart_bridge/uart_bridge.cpp
*
* Usage example for the UART Expansion IO on the Gen3 Interconnect module.
*
* PHYSICAL SETUP:
* =========
* For this example to work, you will have to connect the RX and TX pins on the interconnect 
* IO Expansion pinout, so that the TX is transmitted right into the RX.
*
* DESCRIPTION OF CURRENT EXAMPLE:
* ===============================
* In this example, the UART bridge class encapsulates all necessary Kortex API
* objects and implements the functions to setup the UART, write to the bus and read back from it.
* 
* The Init function creates the Kortex API objects, connects to the arm and also calls InitSocket, 
* which correctly configures the socket (depending if you are on Windows or Linux).
*
* The Run function :
*   - Tells the Interconnect (m_interconnect_config object) to enable the UART bridge (this is where you supply the UART parameters)
*   - Tells the Base (m_base) to enable the UART bridge
*   - Opens the TCP socket to the base and writes the test string to the UART bridge
*   - Waits 10 seconds and prints every character it receives
*   - Tells the Base (m_base) to disable the UART bridge
*   - Tells the Interconenct (m_interconnect_config_object) to disable the UART bridge
*/

#include <BaseClientRpc.h>
#include <InterconnectConfigClientRpc.h>
#include <SessionManager.h>
#include <DeviceManagerClientRpc.h>
#include <RouterClient.h>
#include <TransportClientTcp.h>

#ifdef _WIN32
/* See http://stackoverflow.com/questions/12765743/getaddrinfo-on-win32 */
#include <winsock2.h>
#include <Ws2tcpip.h>
#else
/* Assume that any non-Windows platform uses POSIX-style sockets instead. */
#include <sys/time.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#define SOCKET int
#endif
namespace k_api = Kinova::Api;

#define IP_ADDRESS "192.168.1.10"
#define PORT 10000

///////////////////////////////////////////////////////////////////////
// class UARTBridge
//
// Implements methods for establishing and operating UART bridge through
// the base.
///////////////////////////////////////////////////////////////////////
class UARTBridge
{
public:
    UARTBridge(const std::string& ip_address, int port, const std::string& username = "admin", const std::string& password = "admin"):
    m_ip_address(ip_address), m_username(username), m_password(password), m_port(port), m_interconnect_device_id(0)
    {
        m_router              = nullptr;
        m_transport           = nullptr;
        m_session_manager     = nullptr;
        m_base                = nullptr;
        m_device_manager      = nullptr;
        m_interconnect_config = nullptr;
    }

    ~UARTBridge()
    {
        // Close API session
        m_session_manager->CloseSession();

        // Deactivate the router and cleanly disconnect from the transport object
        m_router->SetActivationStatus(false);
        m_transport->disconnect();

        // Destroy the API
        delete m_interconnect_config;
        delete m_device_manager;
        delete m_session_manager;
      	delete m_base;
        delete m_router;
        delete m_transport;
    }

    void Init()
    {
        m_transport = new k_api::TransportClientTcp();
        m_transport->connect(m_ip_address, m_port);
        m_router = new k_api::RouterClient(m_transport, [](k_api::KError err){ std::cout << "_________ callback error _________" << err.toString(); });
        
        // Set session data connection information
        auto createSessionInfo = k_api::Session::CreateSessionInfo();
        createSessionInfo.set_username(m_username);
        createSessionInfo.set_password(m_password);
        createSessionInfo.set_session_inactivity_timeout(60000);   // (milliseconds)
        createSessionInfo.set_connection_inactivity_timeout(2000); // (milliseconds)

        // Session manager service wrapper
        m_session_manager = new k_api::SessionManager(m_router);
        m_session_manager->CreateSession(createSessionInfo);

        // Create services
        m_device_manager = new k_api::DeviceManager::DeviceManagerClient(m_router);
        m_interconnect_config = new k_api::InterconnectConfig::InterconnectConfigClient(m_router);
        m_base = new k_api::Base::BaseClient(m_router);

        m_interconnect_device_id = GetDeviceIdFromDevType(k_api::Common::INTERCONNECT, 0);
    
        InitSocket();
    }

    uint32_t GetDeviceIdFromDevType(k_api::Common::DeviceTypes device_type, uint32_t device_index)
    {
        auto device_handles = m_device_manager->ReadAllDevices();
        uint32_t current_index = 0;
        for (auto device_handle : device_handles.device_handle() )
        {
            if (device_handle.device_type() == device_type)
            {
                if (current_index == device_index)
                {
                    return device_handle.device_identifier();
                }
                ++current_index;
            }
        }
        return 0;
    }

    void Run()
    {
        const char test_string[] = "This is a string written to the UART using the bridge\n";

        if (m_interconnect_device_id == 0)
        {
            std::cout << "Error : No interconnect found!" << std::endl;
            return;
        }
        std::cout << "Interconnect ID found = " <<  m_interconnect_config  << std::endl;

        ///////////////////////////////////////////////////////////////////////
        // Enable and configure UART on interconnect
        ///////////////////////////////////////////////////////////////////////

        k_api::Common::UARTConfiguration uart_config;
        uart_config.set_port_id(k_api::InterconnectConfig::UART_PORT_EXPANSION);
        uart_config.set_enabled(true);   // Setting enabled to true opens the tcp port dedicated to UART bridging. Setting this
                                        // field to false disables designated uart and closes the TCP port.

        uart_config.set_speed(k_api::Common::UART_SPEED_115200);
        uart_config.set_word_length(k_api::Common::UART_WORD_LENGTH_8);
        uart_config.set_stop_bits(k_api::Common::UART_STOP_BITS_1);
        uart_config.set_parity(k_api::Common::UART_PARITY_NONE);
        m_interconnect_config->SetUARTConfiguration(uart_config, m_interconnect_device_id);

        ///////////////////////////////////////////////////////////////////////
        // Enable port bridging on the base
        ///////////////////////////////////////////////////////////////////////

        k_api::Base::BridgeConfig bridge_config;
        bridge_config.set_device_identifier(m_interconnect_device_id);
        bridge_config.set_bridgetype(k_api::Base::BRIDGE_TYPE_UART);
        auto bridge_result = m_base->EnableBridge(bridge_config);
        if (bridge_result.status() != k_api::Base::BRIDGE_STATUS_OK) 
        {
            return;
        }

        ///////////////////////////////////////////////////////////////////////
        // Get the bridge config and print it
        ///////////////////////////////////////////////////////////////////////

        uint32_t bridge_id = bridge_result.bridge_id().bridge_id();
        k_api::Base::BridgeIdentifier bridge_identifier;
        bridge_identifier.set_bridge_id(bridge_id);
        bridge_config = m_base->GetBridgeConfig(bridge_identifier);

        uint16_t base_port           = (uint16_t) bridge_config.port_config().out_port();
        uint16_t interconnect_port   = (uint16_t) bridge_config.port_config().target_port();

        std::cout << "Bridge ID# " << bridge_id <<" created between Interconnect (dev#" << m_interconnect_device_id << ")";
        std::cout << " port #" << interconnect_port << " and external port #"<< base_port << std::endl;

        ///////////////////////////////////////////////////////////////////////
        // Open TCP socket to bridge port and send sample data through socket
        ///////////////////////////////////////////////////////////////////////

        SOCKET uart_socket = ConnectSocket(IP_ADDRESS, base_port);
        if (uart_socket >= 0)
        {
            send(uart_socket, test_string, sizeof(test_string), 0 );
        }

        ///////////////////////////////////////////////////////////////////////
        // Wait 10 seconds to receive data from socket
        ///////////////////////////////////////////////////////////////////////

        char rx_char;
        std::cout << "Waiting about 10 seconds to receive data from UART" << std::endl;
        std::cout << "Characters received : ";
        std::flush(std::cout);

        // Display received characters from UART bridge socket for 10 seconds.
        auto startTime = std::chrono::system_clock::now();
        auto currentTime = startTime;
        while ( std::chrono::duration<double>(currentTime - startTime).count() < 10.0)
        {
            // Wait for data for about 10 seconds.
            if (GetSocketChar(uart_socket, &rx_char))
            {
                std::cout << rx_char;
                std::flush(std::cout);
            }
            currentTime = std::chrono::system_clock::now();
        }

        std::cout << std::endl;

        std::cout << "Test over!" << std::endl;

        ///////////////////////////////////////////////////////////////////////
        // Close TCP connection and disable bridge on interconnect and base
        ///////////////////////////////////////////////////////////////////////

        CloseSocket(uart_socket);
        
        // Disable bridge on Base
        bridge_identifier.set_bridge_id(bridge_id);
        m_base->DisableBridge(bridge_identifier);

        // Disable UART bridge on Interconnect
        uart_config.set_enabled(false);
        m_interconnect_config->SetUARTConfiguration(uart_config, m_interconnect_device_id);

        QuitSocket();
    }

private:
    k_api::RouterClient*                                 m_router;
    k_api::TransportClientTcp*                           m_transport;
    k_api::SessionManager*                               m_session_manager;
    k_api::Base::BaseClient*                             m_base;
    k_api::DeviceManager::DeviceManagerClient*           m_device_manager;
    k_api::InterconnectConfig::InterconnectConfigClient* m_interconnect_config;

    std::string m_username;
    std::string m_password;
    std::string m_ip_address;
    int         m_port;
    uint32_t    m_interconnect_device_id;

    // Private methods
    int InitSocket(void)
    {
#ifdef _WIN32
        WSADATA wsa_data;
        return WSAStartup(MAKEWORD(1,1), &wsa_data);
#else
        return 0;
#endif
    }

    int QuitSocket(void)
    {
#ifdef _WIN32
        return WSACleanup();
#else
        return 0;
#endif
    }

    int CloseSocket(SOCKET sock)
    {
        int status = 0;
#ifdef _WIN32
        status = shutdown(sock, SD_BOTH);
        if (status == 0)
        {
            status = closesocket(sock);
        }
#else
        status = shutdown(sock, SHUT_RDWR);
        if (status == 0)
        {
            status = close(sock);
        }
#endif
        return status;
    }

    SOCKET ConnectSocket(const char* ip_address, int port)
    {
        SOCKET sock;
        struct sockaddr_in server;
        //Create socket
        sock = socket(AF_INET , SOCK_STREAM , 0);
        if (sock < 0)
        {
            return -1;
        }

        server.sin_addr.s_addr = inet_addr(ip_address);
        server.sin_family = AF_INET;
        server.sin_port = htons(port);
        if (connect(sock , (struct sockaddr *)&server , sizeof(server)) < 0)
        {
            std::cout << "Connect failed." << std::endl;
            return -1;
        }
        return sock;
    }

    bool GetSocketChar(int sock, char* out_char)
    {
        if (out_char == nullptr)
        {
            return false;
        }
        if (sock == 0)
        {
            return false;
        }

        struct timeval ts;
        ts.tv_sec = 0; // 1 millisecond
        ts.tv_usec = 0;
        fd_set fds;
        FD_ZERO(&fds);
        FD_SET(sock, &fds);

        int nready = select(sock + 1, &fds, (fd_set *) 0, (fd_set *) 0, &ts);
        if (nready <= 0)
        {
            return false;
        }
        if (FD_ISSET(sock, &fds))
        {
            if (recv(sock, out_char, 1, 0) > 0)
            {
                return true;
            }
        }
        return false;
    }
};

// Example core
// Implements a sample UART bridge application.
int main(int argc, char **argv)
{
    UARTBridge* uart_bridge;
    uart_bridge = new UARTBridge(IP_ADDRESS, PORT, "admin", "admin");
    uart_bridge->Init();
    uart_bridge->Run();

    delete uart_bridge;

    return 0;
};
